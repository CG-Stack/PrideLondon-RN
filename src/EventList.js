// @flow
import React from "react";
import { View, StyleSheet, FlatList } from "react-native";
import EventCard from "./EventCard";
import type { Event } from "./integrations/cms";

type Props = {
  locale: string,
  events: Event[],
  refreshing: boolean,
  onRefresh: () => void
};

const EventList = ({ events, locale, refreshing, onRefresh }: Props) => (
  <FlatList
    data={events}
    keyExtractor={event => event.sys.id}
    renderItem={({ item: event }) => (
      <View style={styles.eventListItem}>
        <EventCard name={event.fields.name[locale]} />
      </View>
    )}
    refreshing={refreshing}
    onRefresh={onRefresh}
  />
);

const styles = StyleSheet.create({
  eventListItem: {
    paddingLeft: 15,
    paddingRight: 15,
    paddingBottom: 10
  }
});

export default EventList;
