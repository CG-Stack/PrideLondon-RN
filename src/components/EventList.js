// @flow
import React from "react";
import { StyleSheet, SectionList, TouchableOpacity, View } from "react-native";
import type { SectionBase } from "react-native/Libraries/Lists/SectionList";
import formatDate from "date-fns/format";

import EventCard from "./EventCard";
import Text from "./Text";
import type { Event, EventDays, Asset } from "../data/event";
import {
  bgColor,
  eventCardTextColor,
  sectionHeaderShadow,
  sectionHeaderBgColor,
  eventCardShadow
} from "../constants/colors";

type Props = {
  locale: string,
  events: EventDays,
  refreshing?: boolean,
  onRefresh?: () => void,
  onPress: (eventName: string) => void,
  getAssetById: string => Asset
};

const separator = style => () => <View style={style} />;

type ItemProps = { item: Event };
const renderItem = (styles, locale, onPress, getAssetById) => ({
  item: event
}: ItemProps) => (
  <TouchableOpacity
    style={styles.eventListItem}
    accessibilityTraits={["button"]}
    accessibilityComponentType="button"
    delayPressIn={50}
    onPress={() => onPress(event.sys.id)}
  >
    <EventCard
      name={event.fields.name[locale]}
      locationName={event.fields.locationName[locale]}
      eventPriceLow={event.fields.eventPriceLow[locale]}
      eventPriceHigh={event.fields.eventPriceHigh[locale]}
      startTime={event.fields.startTime[locale]}
      endTime={event.fields.endTime[locale]}
      imageUrl={`https:${
        getAssetById(event.fields.eventsListPicture[locale].sys.id).fields.file[
          locale
        ].url
      }`}
      isFree={event.fields.isFree[locale]}
    />
  </TouchableOpacity>
);

type Section = SectionBase<Event> & { title: string };

type SectionProps = { section: Section };
const renderSectionHeader = styles => ({ section }: SectionProps) => (
  <Text type="h2" style={styles.sectionHeader}>
    {section.title}
  </Text>
);

const eventSections = (events: EventDays, locale: string): Section[] =>
  events.map(it => ({
    data: it,
    title: formatDate(it[0].fields.startTime[locale], "dddd D MMMM")
  }));

const EventList = ({
  events,
  locale,
  refreshing,
  onRefresh,
  onPress,
  getAssetById
}: Props) => (
  <SectionList
    stickySectionHeadersEnabled
    sections={eventSections(events, locale)}
    renderSectionHeader={renderSectionHeader(styles)}
    renderSectionFooter={separator(styles.sectionFooter)}
    renderItem={renderItem(styles, locale, onPress, getAssetById)}
    keyExtractor={event => event.sys.id}
    contentContainerStyle={styles.container}
    ItemSeparatorComponent={separator(styles.itemSeparator)}
    SectionSeparatorComponent={separator(styles.sectionSeparator)}
    refreshing={refreshing}
    onRefresh={onRefresh}
  />
);

EventList.defaultProps = {
  refreshing: false,
  onRefresh: undefined
};

const styles = StyleSheet.create({
  itemSeparator: {
    height: 12
  },
  sectionSeparator: {
    height: 6
  },
  container: {
    paddingTop: 0,
    backgroundColor: bgColor
  },
  eventListItem: {
    marginHorizontal: 16,
    borderRadius: 5,
    // The below properties are required for ioS shadow
    shadowColor: eventCardShadow,
    shadowOffset: { width: 0, height: 1 },
    shadowOpacity: 1,
    shadowRadius: 3,
    // The below properties are required for android shadow
    borderWidth: 0,
    elevation: 3,
    backgroundColor: bgColor
  },
  sectionHeader: {
    height: 40,
    color: eventCardTextColor,
    paddingVertical: 8,
    paddingHorizontal: 15,
    backgroundColor: sectionHeaderBgColor,

    // The below properties are required for ioS shadow
    shadowColor: sectionHeaderShadow,
    shadowOffset: { width: 0, height: 1 },
    shadowOpacity: 1,
    shadowRadius: 3,
    // The below properties are required for android shadow
    borderWidth: 0,
    elevation: 3,
    marginBottom: 6
  },
  sectionFooter: {
    height: 6
  }
});

export default EventList;
