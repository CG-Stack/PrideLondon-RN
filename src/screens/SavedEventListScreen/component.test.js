// @flow
import React from "react";
import type { NavigationScreenProp, NavigationState } from "react-navigation";
import { shallow } from "enzyme";
import Component from "./component";
import EventList from "../../components/EventList";

const navigation: NavigationScreenProp<NavigationState> = ({
  navigate: () => {}
}: any);

describe("EventsScreen Component", () => {
  it("renders correctly", () => {
    const output = shallow(
      <Component
        navigation={navigation}
        events={[]}
        loading={false}
        refreshing={false}
        updateEvents={() => Promise.resolve()}
        getAssetUrl={() => ""}
        selectedCategories={new Set()}
        addSavedEvent={() => {}}
        removeSavedEvent={() => {}}
        savedEvents={new Set()}
      />
    );
    expect(output).toMatchSnapshot();
  });

  it("renders loading indicator when loading", () => {
    const output = shallow(
      <Component
        navigation={navigation}
        events={[]}
        loading
        refreshing={false}
        updateEvents={() => Promise.resolve()}
        getAssetUrl={() => ""}
        selectedCategories={new Set()}
        addSavedEvent={() => {}}
        removeSavedEvent={() => {}}
        savedEvents={new Set()}
      />
    );

    const loadingText = output.find("Text").last();

    expect(loadingText.children().text()).toEqual("Loading...");
  });

  it("updates events on refresh", () => {
    const updateEvents = jest.fn();
    const output = shallow(
      <Component
        navigation={navigation}
        events={[]}
        loading={false}
        refreshing={false}
        updateEvents={updateEvents}
        getAssetUrl={() => ""}
        selectedCategories={new Set()}
        addSavedEvent={() => {}}
        removeSavedEvent={() => {}}
        savedEvents={new Set()}
      />
    );

    output
      .find(EventList)
      .props()
      .onRefresh();

    expect(updateEvents).toHaveBeenCalled();
  });
});
