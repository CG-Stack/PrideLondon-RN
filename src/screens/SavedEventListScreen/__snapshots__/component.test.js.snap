// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventsScreen Component renders correctly 1`] = `
<Component
  style={
    Object {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
>
  <Header
    backgroundColor="#2d2f7f"
  >
    <ContentPadding
      padding={Object {}}
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 48,
          "justifyContent": "center",
        }
      }
    >
      <Text
        color="blackColor"
        markdown={false}
        markdownStyle={Object {}}
        style={
          Object {
            "color": "#ffffff",
          }
        }
        type="h2"
      >
        Saved Events
      </Text>
    </ContentPadding>
  </Header>
  <EventList
    addSavedEvent={[Function]}
    events={
      Array [
        Object {
          "name": "super duper event",
        },
      ]
    }
    getAssetUrl={[Function]}
    locale="en-GB"
    onPress={[Function]}
    onRefresh={[Function]}
    refreshing={false}
    removeSavedEvent={[Function]}
    savedEvents={Set {}}
  />
</Component>
`;

exports[`EventsScreen Component renders no saved events screen when there are no events 1`] = `
<Component
  style={
    Object {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
>
  <Header
    backgroundColor="#2d2f7f"
  >
    <ContentPadding
      padding={Object {}}
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 48,
          "justifyContent": "center",
        }
      }
    >
      <Text
        color="blackColor"
        markdown={false}
        markdownStyle={Object {}}
        style={
          Object {
            "color": "#ffffff",
          }
        }
        type="h2"
      >
        Saved Events
      </Text>
    </ContentPadding>
  </Header>
  <NoSavedEvents
    navigation={
      Object {
        "navigate": [Function],
      }
    }
  />
</Component>
`;
